### PR TITLE
Recurring Billing Agreement (RBA) Metadata Support

### PR DESCRIPTION
**Pull request by scannillo**
### Summary

This initiative allows merchants to provide more transparency to their users regarding the details of their Recurring Billing Agreement (RBA) including billing schedule, dates, amounts, free trail details, etc. These details will be displayed in a redesigned XO experience with a streamlined, modernized UI.

### Changes

- Add new public types to collect recurring billing details, to set on `BTPayPalVaultRequest`
- Format new params into appropriate JSON POST format for BT GW

### Checklist

- [x] Added a changelog entry

### Authors
@scannillo

### CODE REVIEW

File: Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingPricing.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        case autoReload = "AUTO_RELOAD"
+    }
+    
+    // MARK: - Private Properties
+    
+    private let pricingModel: PricingModel
+    
+    private let amount: String
+    
+    private let reloadThresholdAmount: String?
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingPricing.swift:21]:
```suggestion
    private let pricingModel: PricingModel
    private let amount: String
    private let reloadThresholdAmount: String?
```

File: Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingPricing.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        self.reloadThresholdAmount = reloadThresholdAmount
+    }
+    
+    // MARK: - Internal Methods
+    
+    func parameters() -> [String: Any] {
+        var parameters: [String: Any] = [:]
+        
+        parameters["pricing_model"] = pricingModel.rawValue
+        parameters["price"] = amount
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingPricing.swift:42]:
```suggestion
        var parameters: [String: Any] = [
              "pricing_model": pricingModel.rawValue,
              "price": amount
        ]
```

File: Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingCycle.swift [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    
+    // MARK: - Internal Methods
+    
+    func parameters() -> [String: Any] {
+        var parameters: [String: Any] = [:]
+        
+        parameters["billing_frequency"] = intervalCount
+        parameters["billing_frequency_unit"] = interval.rawValue
+        parameters["number_of_executions"] = numberOfExecutions
+        parameters["trial"] = isTrial
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalBillingCycle.swift:69]:
```suggestion
        var parameters: [String: Any] = [
            "billing_frequency": intervalCount,
            "billing_frequency_unit": interval.rawValue,
            "number_of_executions": numberOfExecutions,
            "trial": isTrial
         ]
```

File: Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalRecurringBillingDetails.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    
+    private let productQuantity: Int?
+    
+    private let oneTimeFeeAmount: String?
+    
+    private let shippingAmount: String?
+    
+    private let productAmount: String?
+    
+    private let taxAmount: String?
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalRecurringBillingDetails.swift:24]:
```suggestion
    private let billingCycles: [BTPayPalBillingCycle]
    private let currencyISOCode: String
    private let productName: String?
    private let productDescription: String?
    private let productQuantity: Int?
    private let oneTimeFeeAmount: String?
    private let shippingAmount: String?
    private let productAmount: String?
    private let taxAmount: String?
```

File: Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalRecurringBillingDetails.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        self.taxAmount = taxAmount
+    }
+    
+    // MARK: - Internal Methods
+    
+    func parameters() -> [String: Any] {
+        var parameters: [String: Any] = [:]
+        
+        parameters["currency_iso_code"] = currencyISOCode
+        parameters["billing_cycles"] = billingCycles.map({ $0.parameters() })
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/RecurringBillingMetadata/BTPayPalRecurringBillingDetails.swift:67]:
```suggestion
        var parameters: [String: Any] = [
            "currency_iso_code": currencyISOCode,
            "billing_cycles": billingCycles.map { $0.parameters() }
        ]
```

File: Demo/Application/Features/PayPalWebCheckoutViewController.swift [hunk 6]
```diff
@@ -55,7 +64,11 @@ class PayPalWebCheckoutViewController: PaymentButtonBaseViewController {
             UIStackView(arrangedSubviews: [newPayPalCheckoutToggleLabel, newPayPalCheckoutToggle]),
             payPalCheckoutButton
         ])
-        let vaultStackView = buttonsStackView(label: "Vault",views: [payPalVaultButton, payPalAppSwitchButton])
+        let vaultStackView = buttonsStackView(label: "Vault", views: [
```
[the remark this answers is not in the document]
Comment (jaxdesmarais) [reply Demo/Application/Features/PayPalWebCheckoutViewController.swift:67]: 🤔 odd, looking at the `files` tab for this PR I see the following:
Screenshot 2024-06-25 at 1 30 06 PM

File: Demo/Application/Features/PayPalWebCheckoutViewController.swift [hunk 7]
```diff
@@ -40,6 +40,15 @@ class PayPalWebCheckoutViewController: PaymentButtonBaseViewController {
...
+        let label = UILabel()
+        label.text = "Recurring Billing (RBA) Data"
+        label.font = .preferredFont(forTextStyle: .footnote)
+        return label
+    }()
+    
+    let rbaDataToggle = UISwitch()
 
     override func viewDidLoad() {
         super.heightConstraint = 300
```
Comment (jaxdesmarais) [inline Demo/Application/Features/PayPalWebCheckoutViewController.swift:54]: Looks like the view currently is a bit squished, if we update this it should give the checkout view a bit more space:
```suggestion
        super.heightConstraint = 350
```

**Review by jaxdesmarais — APPROVED** [review]

**Review by richherrera — APPROVED** [review]

**Pull request merged**